### PR TITLE
Randomized Block Krylov Method.

### PR DESCRIPTION
**Pull request by zoq**
Implementation of the randomized block krylov svd method from the "Randomized Block Krylov Methods for Stronger and Faster Approximate Singular Value Decomposition" paper.

Turns out the randomized svd method causes some instabilities when used for the edges boxes method  since the exact svd method takes way to much time the randomized block krylov svd method is a good alternative. Note the method gives more accurate results for fewer iterations but is slower per iteration than the randomized svd method. Here are some runtime timings for different datasets.

```
data size: 10x10
----------------------------
svd: 0.399944 ms
econ: 0.212553 ms -- error: 0
krylov: 0.675249 ms -- error: 3.84883e-16
randomized: 0.740294 ms -- error: 1.73457e-08

data size: 100x100
----------------------------
svd: 15.5593 ms
econ: 18.658 ms -- error: 0
krylov: 11.7311 ms -- error: 6.52892e-16
randomized: 9.72015 ms -- error: 7.35058e-09

data size: 1000x1000
----------------------------
svd: 615.481 ms
econ: 619.679 ms -- error: 0
krylov: 169.624 ms -- error: 0.0204925
randomized: 60.7702 ms -- error: 0.0603914

data size: 5000x5000
----------------------------
svd: 69463.5 ms
econ: 67744.6 ms -- error: 0
krylov: 1405.95 ms -- error: 0.0510657
randomized: 827.77 ms -- error: 0.0966308

data size: 4000x8000
----------------------------
svd: 72039.1 ms
econ: 57362.5 ms -- error: 0
krylov: 2108.01 ms -- error: 0.0587171
randomized: 1299.3 ms -- error: 0.0979414

data size: 6000x12000
----------------------------
svd: 223858 ms
econ: 176574 ms -- error: 0
krylov: 3975.92 ms -- error: 0.0654123
randomized: 2541.26 ms -- error: 0.104083

data size: 7000x19000
----------------------------
svd: 526310 ms
econ: 332015 ms -- error: 0
krylov: 7915.1 ms -- error: 0.0662568
randomized: 4828.26 ms -- error: 0.107567
```

test code:

```
void test(size_t rows, size_t cols, size_t rank)
{
  arma::mat data;
  data = arma::randn<arma::mat>(rows, cols);

  cout << "\ndata size: " << data.n_rows << "x" << data.n_cols << endl;
  cout << "----------------------------\n";

  arma::mat U1, U2, U3, U4, U5;
  arma::mat V1, V2, V3, V4, V5;
  arma::vec s1, s2, s3, s4, s5;

  auto start = chrono::steady_clock::now();
  arma::svd(U1, s1, V1, data);
  auto end = chrono::steady_clock::now();
  auto diff = end - start;
  cout << "svd: " << chrono::duration <double, milli> (diff).count() << " ms" << endl;

  start = chrono::steady_clock::now();
  arma::svd_econ(U2, s2, V2, data);
  end = chrono::steady_clock::now();
  diff = end - start;

  double error = arma::norm(s2.subvec(0, rank - 1) - s1.subvec(
      0, rank - 1), "frob") / arma::norm(s2.subvec(0, rank - 1), "frob");
  cout << "econ: " << chrono::duration <double, milli> (diff).count() << " ms -- error: "
      << error << endl;

  start = chrono::steady_clock::now();
  svd::RandomizedBlockKrylovSVD bSVD(data, U3, s3, V3, 2, rank, rank);
  end = chrono::steady_clock::now();
  diff = end - start;
  error = arma::norm(s3.subvec(0, rank - 1) - s1.subvec(
      0, rank - 1), "frob") / arma::norm(s3.subvec(0, rank - 1), "frob");
  cout << "krylov: " << chrono::duration <double, milli> (diff).count() << " ms -- error: "
      << error << endl;

  start = chrono::steady_clock::now();
  svd::RandomizedSVD rSVD(data, U4, s4, V4, 0, 2, rank);
  end = chrono::steady_clock::now();
  diff = end - start;
    error = arma::norm(s4.subvec(0, rank - 1) - s1.subvec(
      0, rank - 1), "frob") / arma::norm(s4.subvec(0, rank - 1), "frob");
  cout << "randomized: " << chrono::duration <double, milli> (diff).count() << " ms -- error: "
      << error << endl;
}

BOOST_AUTO_TEST_CASE(CompareSVDTest)
{
  test(10, 10, 10);
  test(100, 100, 100);
  test(1000, 1000, 100);
  test(5000, 5000, 100);
  test(4000, 8000, 100);
  test(6000, 12000, 100);
  test(7000, 19000, 100);
}
```

### CODE REVIEW

File: src/mlpack/tests/block_krylov_svd_test.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+  // The relative reconstruction error should be small.
+  error = arma::norm(centeredData - reconstruct, "frob") /
+      arma::norm(centeredData, "frob");
+  BOOST_REQUIRE_SMALL(error, 1e-5);
+}
+
+/*
+ * Check if the method can handle noisy matrices.
+ */
+BOOST_AUTO_TEST_CASE(RandomizedBlockKrylovSVDNosiyLowRankTest)
```
Comment (rcurtin) [inline src/mlpack/tests/block_krylov_svd_test.cpp]: Nosiy -> Noisy, small typo :)

File: src/mlpack/methods/block_krylov_svd/randomized_block_krylov_svd.cpp [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+  if (blockSize == 0)
+  {
+    blockSize = rank + 10;
+  }
+
+  // Random block initialization.
+  arma::mat G = arma::randn(data.n_rows, blockSize);
+
+  // Construct and orthonormalize Krlov subspace.
```
Comment (rcurtin) [inline src/mlpack/methods/block_krylov_svd/randomized_block_krylov_svd.cpp]: Looks like a small typo, should be "Krylov".

**Comment by zoq** [comment]
Thanks for taking a closer look at the code. About the copy on line 73, I thought about it and could't we use `K` as auxiliary memory.

```cpp
  arma::mat Q, R, block, blockIteration;

  if (blockSize == 0)
  {
    blockSize = rank + 10;
  }

  // Random block initialization.
  arma::mat G = arma::randn(data.n_rows, blockSize);

  // Construct and orthonormalize Krlov subspace.
  arma::mat K(data.n_rows, blockSize * (maxIterations + 1));
  block = arma::mat(K.memptr(), data.n_rows, blockSize, false);

  arma::qr_econ(block, R, data * G);
  for (size_t i = 0, b = block.n_cols, blockOffset = block.n_elem;
      i < maxIterations; ++i, b += block.n_cols, blockOffset += block.n_elem)
  {
    blockIteration = arma::mat(K.memptr() + blockOffset, data.n_rows,
        blockSize, false);

    arma::qr_econ(blockIteration, R, data * (data.t() * block));

    block = arma::mat(K.memptr() + blockOffset, data.n_rows, blockSize, false);
  }
```

**Comment by rcurtin** [comment]
Ah, clever!  That could definitely work.

**Comment by rcurtin** [comment]
No more comments from my end, feel free to merge whenever you think it is ready.  I do think it would be nice to add support in the `mlpack_pca` program but I think I mentioned that already. :)

**Comment by zoq** [comment]
I'll make a separate commit which incorporates the addition into the PCA method.

**Pull request merged**